README.md
<div align="center">
    <img src="https://apps.soldat2d.com/projects/libdb/logo/48x48.png">
</div>

# libdb

## About

**libdb** is an external library for [Soldat](https://github.com/Soldat/soldat) dedicated servers (Windows and Linux) to access SQLite, MySQL/MariaDB, PostgreSQL or any database type (Oracle, Access, etc) via ODBC.

## Usage

An attached archive from the releases section has the following structure:

    libdb-0.3.dll
    libdb-0.3.so
    libmariadb.dll
    libmariadb.so
    scripts
    ├── libdb
    │   ├── database.pas
    │   └── libdb.pas
    └── libdb_test
        ├── config.ini
        └── main.pas

Where `libdb.pas` is a unit with external functions bindings, `database.pas` is a unit with some helper functions and `libdb_test` folder contains an example script that demonstrates the use of `libdb` with SQLite, MySQL/MariaDB and PostgreSQL.

You should extract an archive into the root directory of your Soldat server. Thus `*.dll` and `*.so` files will be placed near to `soldatserver.exe` (Windows) or `soldatserver` (Linux) executables and the rest of the files will be copied into the `./scripts` directory of your server.

Make sure you have `AllowDlls` enabled in the [server.ini](https://wiki.soldat.pl/index.php/SC3_Config_File).

## Notes

### Windows

Everything should work out of the box.

### Linux

`libdb-0.3` depends on `unixODBC`.
For Debian-based distros you could use the following command to install dependencies:

```bash
sudo apt install libodbc1:i386 libltdl7:i386
```

There could be a need to add soldatserver's directory into the `LD_LIBRARY_PATH` environment variable to use external library functions with Soldat server. There are a lot ways of doing it, so here are few of simple ones:

-
    ```bash
    LD_LIBRARY_PATH="$PWD" ./soldatserver
    ```

    this will set current directory (for this particular run) as `LD_LIBRARY_PATH` and run the `soldatserver`

-
    ```bash
    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/path/to/soldatserver && ./soldatserver
    ```

    this will add current directory to `LD_LIBRARY_PATH` and run the `soldatserver`

## Changelog

See [CHANGELOG.md](CHANGELOG.md) for details.

## Discussion

See the post on [forums.soldat.pl](https://forums.soldat.pl/index.php?topic=42012.0).


CHANGELOG.md
 Version | Notes
-------- | -----------
| v0.3   | **27 Jan 2019**
|        | + Added MySQL support (see `DB_Plugin_MySQL`)
|        | + Added SQL bind variables (see `DB_SetString()`, `DB_SetLong()` etc.)
|        | % Modified library bindings and example code (rewritten with `ScriptCore3` syntax)
|        | % Modified some bindings and internal code which broke backward compatibility (see release notes)
| v0.2   | **12 Dec 2012**
|        | % Updated the lib, so its Linux version depends on `unixODBC` instead of `iODBC` now
| v0.1   | **06 Jun 2012**
|        | * First public release

> Note: `+` Added feature, `-` Removed feature, `~` Fixed feature, `%` Modified feature, `*` Comment


scripts/libdb_test/main.pas
uses database;

procedure test_sqlite(FilePath: String);
var
    {$IFDEF INSERT}
    i: Integer;
    {$ENDIF}
    DBFile: TFileStream;
begin
    if File.CheckAccess(FilePath) then
        if File.Exists(FilePath) then begin
            WriteLn('[*] Found: '+FilePath)
        end else begin
            WriteLn('[!] Not found: '+FilePath);
            try
                DBFile := File.CreateFileStream;
                DBFile.SaveToFile(FilePath);
                DBFile.Free;
                WriteLn('[*] Created: '+FilePath);
            except
                WriteLn('Exception: '+ExceptionParam);
                exit;
            end;
        end
    else begin
        WriteLn('Access denied: '+FilePath);
        exit;
    end;

    if DatabaseOpen(DB_Any, FilePath, '', '', DB_Plugin_SQLite) then begin
        if DatabaseQuery(DatabaseID, 'SELECT sqlite_version()') and DB_NextRow(DatabaseID) then begin
            WriteLn('[SQLite v'+DB_GetString(DatabaseID, 0)+ ']');
            DB_FinishQuery(DatabaseID);
        end;

        if DatabaseUpdate(
            DatabaseID,
            'CREATE TABLE IF NOT EXISTS test( '+
                'id INTEGER PRIMARY KEY AUTOINCREMENT, '+
                'name TEXT, '+
                'number INTEGER, '+
                'ratio FLOAT '+
            ');'
        ) then begin
            {$IFDEF INSERT}
            DB_Update(DatabaseID, 'BEGIN;');

            for i := 1 to 3 do begin
                DB_SetString(DatabaseID, 0, '''Name''#'+IntToStr(i));
                DB_SetLong(DatabaseID, 1, i * 10);
                DB_SetFloat(DatabaseID, 2, i / 10.0);

                DatabaseUpdate(DatabaseID, 'INSERT INTO test(name, number, ratio) VALUES(?, ?, ?);');
            end;

            DB_Update(DatabaseID, 'COMMIT;');
            {$ENDIF}

            if DatabaseQuery(DatabaseID, 'SELECT * FROM test;') then begin
                DatabasePrintValues(DatabaseID);
                DB_FinishQuery(DatabaseID);
            end;
        end;

        DB_Close(DatabaseID);
    end;
end;

procedure test_mysql(ConnectionString, User, Password: String);
{$IFDEF INSERT}
var
    i: Integer;
{$ENDIF}
begin
    if DatabaseOpen(DB_Any, ConnectionString, User, Password, DB_Plugin_MySQL) then begin
        if DatabaseQuery(DatabaseID, 'SELECT @@version;') and DB_NextRow(DatabaseID) then begin
            WriteLn('[MySQL v'+DB_GetString(DatabaseID, 0)+ ']');
            DB_FinishQuery(DatabaseID);
        end;

        if DatabaseUpdate(
            DatabaseID,
            'CREATE TABLE IF NOT EXISTS test( '+
                'id INTEGER PRIMARY KEY AUTO_INCREMENT, '+
                'name VARCHAR(25), '+
                'number INTEGER, '+
                'ratio FLOAT '+
            ');'
        ) then begin
            {$IFDEF INSERT}
            DB_Update(DatabaseID, 'START TRANSACTION;');

            for i := 1 to 3 do begin
                DB_SetString(DatabaseID, 0, '''Name''#'+IntToStr(i));
                DB_SetLong(DatabaseID, 1, i * 10);
                DB_SetFloat(DatabaseID, 2, i / 10.0);

                DatabaseUpdate(DatabaseID, 'INSERT INTO test(name, number, ratio) VALUES(?, ?, ?);');
            end;

            DB_Update(DatabaseID, 'COMMIT;');
            {$ENDIF}

            if DatabaseQuery(DatabaseID, 'SELECT * FROM test;') then begin
                DatabasePrintValues(DatabaseID);
                DB_FinishQuery(DatabaseID);
            end;
        end;

        DB_Close(DatabaseID);
    end;
end;

procedure test_postgresql(ConnectionString, User, Password: String);
{$IFDEF INSERT}
var
    i: Integer;
{$ENDIF}
begin
    if DatabaseOpen(DB_Any, ConnectionString, User, Password, DB_Plugin_PostgreSQL) then begin
        if DatabaseQuery(DatabaseID, 'SHOW server_version;') and DB_NextRow(DatabaseID) then begin
            WriteLn('[PostgreSQL v'+DB_GetString(DatabaseID, 0)+ ']');
            DB_FinishQuery(DatabaseID);
        end;

        if DatabaseUpdate(
            DatabaseID,
            'CREATE TABLE IF NOT EXISTS test( '+
                'id SERIAL PRIMARY KEY, '+
                'name VARCHAR(25), '+
                'number INTEGER, '+
                'ratio FLOAT '+
            ');'
        ) then begin
            {$IFDEF INSERT}
            DB_Update(DatabaseID, 'BEGIN;');

            for i := 1 to 3 do begin
                DB_SetString(DatabaseID, 0, '''Name''#'+IntToStr(i));
                DB_SetLong(DatabaseID, 1, i * 10);
                DB_SetFloat(DatabaseID, 2, i / 10.0);

                DatabaseUpdate(DatabaseID, 'INSERT INTO test(name, number, ratio) VALUES($1, $2, $3);');
            end;

            DB_Update(DatabaseID, 'COMMIT;');
            {$ENDIF}

            if DatabaseQuery(DatabaseID, 'SELECT * FROM test;') then begin
                DatabasePrintValues(DatabaseID);
                DB_FinishQuery(DatabaseID);
            end;
        end;

        DB_Close(DatabaseID);
    end;
end;

begin
    test_sqlite('~/test.db');
    test_mysql('host=localhost port=3306 dbname=test', 'user', 'password');
    test_postgresql('host=localhost port=5432 dbname=test', 'user', 'password');
end.


scripts/libdb-0.2/libdb_linux.pas
// =====================================
//          C O N S T A N T S
// =====================================
{ Database plugins enumeration
  for the DB_Open() function   }
Const DB_Plugin_ODBC       = 1;
Const DB_Plugin_SQLite     = 2;
Const DB_Plugin_PostgreSQL = 3;

{ Database column types enumeration 
  for the DB_ColumnType() function  }
Const DB_Type_Double = 1;
Const DB_Type_Float  = 2;
Const DB_Type_Long   = 3;
Const DB_Type_String = 4;

// =====================================
//        D E C L A R A T I O N S
// =====================================
Procedure DB_Close(DatabaseID: Integer);
External 'DB_Close@libdb-0.2.so cdecl';

Function DB_ColumnName(DatabaseID, Column: Integer): PChar;
External 'DB_ColumnName@libdb-0.2.so cdecl';

Function DB_ColumnSize(DatabaseID, Column: Integer): Integer;
External 'DB_ColumnSize@libdb-0.2.so cdecl';

Function DB_ColumnType(DatabaseID, Column: Integer): Integer;
External 'DB_ColumnType@libdb-0.2.so cdecl';

Function DB_Columns(DatabaseID: Integer): Integer;
External 'DB_Columns@libdb-0.2.so cdecl';

Function DB_Error(): PChar;
External 'DB_Error@libdb-0.2.so cdecl';

Function DB_Query(DatabaseID: Integer; Query: PChar): Integer;
External 'DB_Query@libdb-0.2.so cdecl';

Function DB_Update(DatabaseID: Integer; Query: PChar): Integer;
External 'DB_Update@libdb-0.2.so cdecl';

Procedure DB_FinishQuery(DatabaseID: Integer);
External 'DB_FinishQuery@libdb-0.2.so cdecl';

Function DB_FirstRow(DatabaseID: Integer): Integer;
External 'DB_FirstRow@libdb-0.2.so cdecl';

Function DB_GetDouble(DatabaseID, Column: Integer): Double;
External 'DB_GetDouble@libdb-0.2.so cdecl';

Function DB_GetFloat(DatabaseID, Column: Integer): Single;
External 'DB_GetFloat@libdb-0.2.so cdecl';

Function DB_GetLong(DatabaseID, Column: Integer): LongInt;
External 'DB_GetLong@libdb-0.2.so cdecl';

Function DB_GetString(DatabaseID, Column: Integer): PChar;
External 'DB_GetString@libdb-0.2.so cdecl';

Function DB_IsDatabase(DatabaseID: Integer): Integer;
External 'DB_IsDatabase@libdb-0.2.so cdecl';

Function DB_NextRow(DatabaseID: Integer): Integer;
External 'DB_NextRow@libdb-0.2.so cdecl';

Function DB_Open(DatabaseID: Integer; DatabaseName, User, Password: PChar; Plugin: Integer): Integer;
External 'DB_Open@libdb-0.2.so cdecl';

Function DB_ExamineDrivers(): Integer;
External 'DB_ExamineDrivers@libdb-0.2.so cdecl';

Function DB_NextDriver(): Integer;
External 'DB_NextDriver@libdb-0.2.so cdecl';

Function DB_DriverDescription(): PChar;
External 'DB_DriverDescription@libdb-0.2.so cdecl';

Function DB_DriverName(): PChar;
External 'DB_DriverName@libdb-0.2.so cdecl';

Function DB_GetVersion(): Integer;
External 'DB_GetVersion@libdb-0.2.so cdecl';

// =====================================
//            E X A M P L E S
// =====================================
function CheckDatabaseUpdate(Database: Integer; Query: PChar): Integer;
var
	Check: Integer;
begin
	WriteLn('>'+Query); //debug output
	Check:= DB_Update(Database, Query);
	
	if Check = 0 then
		WriteLn(DB_Error());
	
	Result:= Check;
end;


procedure create_db1();
var
  db_path: String;
begin
	db_path:= 'test.db';
	//creating a new file if not exists
	if Not FileExists(db_path) then
		if WriteFile(db_path,'') then
			WriteLn('File "'+db_path+'" created...');
	
	if DB_Open(0, db_path, '', '', DB_Plugin_SQLite) <> 0 then
	begin
		WriteLn('Database "'+db_path+'" opened...');
		CheckDatabaseUpdate(0, 'CREATE TABLE IF NOT EXISTS test(id INTEGER PRIMARY KEY, name STRING, value INTEGER);');
		CheckDatabaseUpdate(0, 'INSERT INTO test(name, value) VALUES("name1", 5);');
		CheckDatabaseUpdate(0, 'INSERT INTO test(name, value) VALUES("name2", 10);');
		CheckDatabaseUpdate(0, 'INSERT INTO test(name, value) VALUES(''name3'', 15);');
		WriteLn('Values inserted...');
		DB_Close(0);
		WriteLn('Database "'+db_path+'" closed...');
	end;
end;


procedure create_db2();
var
  db_path: String;
  query: String;
  db, i: Integer;
begin
	db_path:= 'test.db';
	//creating a new file if not exists
	if (Not FileExists(db_path)) then
		if (WriteFile(db_path,'')) then
			WriteLn('File "'+db_path+'" created...');
	
	db:= 0;
	if DB_Open(db, db_path, '', '', DB_Plugin_SQLite) <> 0 then
	begin
		WriteLn('Database "'+db_path+'" opened...');
		query:= 'CREATE TABLE IF NOT EXISTS test(';
		query:= query+'id INTEGER PRIMARY KEY, ';
		query:= query+'name TEXT NOT NULL, ';
		query:= query+'value INTEGER, ';
		query:= query+'UNIQUE(name) ON CONFLICT IGNORE);';
		if CheckDatabaseUpdate(db, query) <> 0 then
		begin
			DB_Update(db, 'BEGIN;');
			for i:= 1 to 20 do
				CheckDatabaseUpdate(db, 'INSERT INTO test(name, value) VALUES("name'+IntToStr(i)+'", '+IntToStr(i)+');');

			DB_Update(db, 'COMMIT;');
			WriteLn('Values inserted...');
		end;
		DB_Close(db);
		WriteLn('Database "'+db_path+'" closed...');
	end;
end;


procedure query_db(param: Integer);
var
  db_path, query, result_srt: String;
  i, columns: Integer;
begin
	db_path:= 'test.db';
	if DB_Open(0, db_path, '', '', DB_Plugin_SQLite) <> 0 then
	begin
		WriteLn('Database "'+db_path+'" opened...');
		query:= 'SELECT * FROM test WHERE value > '+IntToStr(param)+';';
		if DB_Query(0, query) <> 0 then
		begin
			WriteLn('>'+query); //debug output
			columns:= DB_Columns(0);
			while DB_NextRow(0) <> 0 do
			begin
				result_srt:='';
				for i:= 0 to columns - 1 do
					result_srt:= result_srt +DB_ColumnName(0, i)+': '+DB_GetString(0, i)+'; ';
				
				WriteLn(result_srt);
			end;
			
			DB_FinishQuery(0);
		end;
	
		DB_Close(0);
		WriteLn('Database "'+db_path+'" closed...');
	end
	else
		WriteLn('Can not open database "'+db_path+'"!');
end;


procedure drivers();
begin
	if DB_ExamineDrivers() <> 0 then
	begin
		WriteLn('ODBC drivers installed:');
		while DB_NextDriver() <> 0 do
		begin
			WriteLn('Name - '+DB_DriverName());
			WriteLn('Desc - '+DB_DriverDescription());
			WriteLn('* * * * *');
		end;
	end
	else
		WriteLn('No ODBC drivers installed!');
end;


procedure OnAdminMessage(IP, Msg: String);
begin
	case LowerCase(Msg) of
	'/create1' : create_db1();
	'/create2' : create_db2();
	'/query1'  : query_db(7);
	'/query2'  : query_db(14);
	'/drivers' : drivers();
	'/version' : WriteLn('libdb v0.'+IntToStr(DB_GetVersion()));
	end;
end;


scripts/libdb-0.2/libdb_windows.pas
// =====================================
//          C O N S T A N T S
// =====================================
{ Database plugins enumeration
  for the DB_Open() function   }
Const DB_Plugin_ODBC       = 1;
Const DB_Plugin_SQLite     = 2;
Const DB_Plugin_PostgreSQL = 3;

{ Database column types enumeration 
  for the DB_ColumnType() function  }
Const DB_Type_Double = 1;
Const DB_Type_Float  = 2;
Const DB_Type_Long   = 3;
Const DB_Type_String = 4;

// =====================================
//        D E C L A R A T I O N S
// =====================================
Procedure DB_Close(DatabaseID: Integer);
External 'DB_Close@libdb-0.2.dll cdecl';

Function DB_ColumnName(DatabaseID, Column: Integer): PChar;
External 'DB_ColumnName@libdb-0.2.dll cdecl';

Function DB_ColumnSize(DatabaseID, Column: Integer): Integer;
External 'DB_ColumnSize@libdb-0.2.dll cdecl';

Function DB_ColumnType(DatabaseID, Column: Integer): Integer;
External 'DB_ColumnType@libdb-0.2.dll cdecl';

Function DB_Columns(DatabaseID: Integer): Integer;
External 'DB_Columns@libdb-0.2.dll cdecl';

Function DB_Error(): PChar;
External 'DB_Error@libdb-0.2.dll cdecl';

Function DB_Query(DatabaseID: Integer; Query: PChar): Integer;
External 'DB_Query@libdb-0.2.dll cdecl';

Function DB_Update(DatabaseID: Integer; Query: PChar): Integer;
External 'DB_Update@libdb-0.2.dll cdecl';

Procedure DB_FinishQuery(DatabaseID: Integer);
External 'DB_FinishQuery@libdb-0.2.dll cdecl';

Function DB_FirstRow(DatabaseID: Integer): Integer;
External 'DB_FirstRow@libdb-0.2.dll cdecl';

Function DB_GetDouble(DatabaseID, Column: Integer): Double;
External 'DB_GetDouble@libdb-0.2.dll cdecl';

Function DB_GetFloat(DatabaseID, Column: Integer): Single;
External 'DB_GetFloat@libdb-0.2.dll cdecl';

Function DB_GetLong(DatabaseID, Column: Integer): LongInt;
External 'DB_GetLong@libdb-0.2.dll cdecl';

Function DB_GetString(DatabaseID, Column: Integer): PChar;
External 'DB_GetString@libdb-0.2.dll cdecl';

Function DB_IsDatabase(DatabaseID: Integer): Integer;
External 'DB_IsDatabase@libdb-0.2.dll cdecl';

Function DB_NextRow(DatabaseID: Integer): Integer;
External 'DB_NextRow@libdb-0.2.dll cdecl';

Function DB_Open(DatabaseID: Integer; DatabaseName, User, Password: PChar; Plugin: Integer): Integer;
External 'DB_Open@libdb-0.2.dll cdecl';

Function DB_ExamineDrivers(): Integer;
External 'DB_ExamineDrivers@libdb-0.2.dll cdecl';

Function DB_NextDriver(): Integer;
External 'DB_NextDriver@libdb-0.2.dll cdecl';

Function DB_DriverDescription(): PChar;
External 'DB_DriverDescription@libdb-0.2.dll cdecl';

Function DB_DriverName(): PChar;
External 'DB_DriverName@libdb-0.2.dll cdecl';

Function DB_GetVersion(): Integer;
External 'DB_GetVersion@libdb-0.2.dll cdecl';

// =====================================
//            E X A M P L E S
// =====================================
function CheckDatabaseUpdate(Database: Integer; Query: PChar): Integer;
var
	Check: Integer;
begin
	WriteLn('>'+Query); //debug output
	Check:= DB_Update(Database, Query);
	
	if Check = 0 then
		WriteLn(DB_Error());
	
	Result:= Check;
end;


procedure create_db1();
var
  db_path: String;
begin
	db_path:= 'test.db';
	//creating a new file if not exists
	if Not FileExists(db_path) then
		if WriteFile(db_path,'') then
			WriteLn('File "'+db_path+'" created...');
	
	if DB_Open(0, db_path, '', '', DB_Plugin_SQLite) <> 0 then
	begin
		WriteLn('Database "'+db_path+'" opened...');
		CheckDatabaseUpdate(0, 'CREATE TABLE IF NOT EXISTS test(id INTEGER PRIMARY KEY, name STRING, value INTEGER);');
		CheckDatabaseUpdate(0, 'INSERT INTO test(name, value) VALUES("name1", 5);');
		CheckDatabaseUpdate(0, 'INSERT INTO test(name, value) VALUES("name2", 10);');
		CheckDatabaseUpdate(0, 'INSERT INTO test(name, value) VALUES(''name3'', 15);');
		WriteLn('Values inserted...');
		DB_Close(0);
		WriteLn('Database "'+db_path+'" closed...');
	end;
end;


procedure create_db2();
var
  db_path: String;
  query: String;
  db, i: Integer;
begin
	db_path:= 'test.db';
	//creating a new file if not exists
	if (Not FileExists(db_path)) then
		if (WriteFile(db_path,'')) then
			WriteLn('File "'+db_path+'" created...');
	
	db:= 0;
	if DB_Open(db, db_path, '', '', DB_Plugin_SQLite) <> 0 then
	begin
		WriteLn('Database "'+db_path+'" opened...');
		query:= 'CREATE TABLE IF NOT EXISTS test(';
		query:= query+'id INTEGER PRIMARY KEY, ';
		query:= query+'name TEXT NOT NULL, ';
		query:= query+'value INTEGER, ';
		query:= query+'UNIQUE(name) ON CONFLICT IGNORE);';
		if CheckDatabaseUpdate(db, query) <> 0 then
		begin
			DB_Update(db, 'BEGIN;');
			for i:= 1 to 20 do
				CheckDatabaseUpdate(db, 'INSERT INTO test(name, value) VALUES("name'+IntToStr(i)+'", '+IntToStr(i)+');');

			DB_Update(db, 'COMMIT;');
			WriteLn('Values inserted...');
		end;
		DB_Close(db);
		WriteLn('Database "'+db_path+'" closed...');
	end;
end;


procedure query_db(param: Integer);
var
  db_path, query, result_srt: String;
  i, columns: Integer;
begin
	db_path:= 'test.db';
	if DB_Open(0, db_path, '', '', DB_Plugin_SQLite) <> 0 then
	begin
		WriteLn('Database "'+db_path+'" opened...');
		query:= 'SELECT * FROM test WHERE value > '+IntToStr(param)+';';
		if DB_Query(0, query) <> 0 then
		begin
			WriteLn('>'+query); //debug output
			columns:= DB_Columns(0);
			while DB_NextRow(0) <> 0 do
			begin
				result_srt:='';
				for i:= 0 to columns - 1 do
					result_srt:= result_srt +DB_ColumnName(0, i)+': '+DB_GetString(0, i)+'; ';
				
				WriteLn(result_srt);
			end;
			
			DB_FinishQuery(0);
		end;
	
		DB_Close(0);
		WriteLn('Database "'+db_path+'" closed...');
	end
	else
		WriteLn('Can not open database "'+db_path+'"!');
end;


procedure drivers();
begin
	if DB_ExamineDrivers() <> 0 then
	begin
		WriteLn('ODBC drivers installed:');
		while DB_NextDriver() <> 0 do
		begin
			WriteLn('Name - '+DB_DriverName());
			WriteLn('Desc - '+DB_DriverDescription());
			WriteLn('* * * * *');
		end;
	end
	else
		WriteLn('No ODBC drivers installed!');
end;


procedure OnAdminMessage(IP, Msg: String);
begin
	case LowerCase(Msg) of
	'/create1' : create_db1();
	'/create2' : create_db2();
	'/query1'  : query_db(7);
	'/query2'  : query_db(14);
	'/drivers' : drivers();
	'/version' : WriteLn('libdb v0.'+IntToStr(DB_GetVersion()));
	end;
end;
